### PR TITLE
Feature/35183/oxford handler

### PR DESCRIPTION
**Pull request by lauramccluskey1**
Details on this change can be found on ticket 35183 and include:

- Extra genes added
- Updates to testscope regular expressions
- Updates to codingdnasequencechange regular expressions
- Functionality to split the files into brca and colorectal

Variant counts have been checked as part of QA process.

### CODE REVIEW

File: lib/import/brca/providers/oxford/oxford_handler.rb [hunk 1]
```diff
@@ -23,6 +25,29 @@ def process_fields(record)
             res&.each { |cur_genotype| @persister.integrate_and_store(cur_genotype) }
           end
 
+          def brca_file?
+            file_name = @batch.original_filename
+            file_path = file_name.slice(0, file_name.rindex('/'))
```
Comment (ollietulloch) [inline lib/import/brca/providers/oxford/oxford_handler.rb:30]: Could you use `File.dirname(@batch.original_filename)` here?

Comment (lauramccluskey1) [reply lib/import/brca/providers/oxford/oxford_handler.rb:30]: file_name variable (line 28) is still required for line 31. So do you want me to just replace line 30 with file_path=File.dirname(file_name)?

Comment (ollietulloch) [reply lib/import/brca/providers/oxford/oxford_handler.rb:30]: You could still assign `file_name = @batch.original_filename` and then do `File.dirname(file_name)`. It's the `file_name.slice(0, file_name.rindex('/'))` that I don't think we need, the `File` class has useful methods for dealing with file names/paths/directories.

Comment (lauramccluskey1) [reply lib/import/brca/providers/oxford/oxford_handler.rb]: @ollietulloch  Changes made and tested- see commit [d4d9719](https://github.com/NHSDigital/data_management_system/pull/56/commits/d4d97193ceab1a562b4be9cd728554a5b8e3526b)

File: lib/import/brca/providers/oxford/oxford_handler.rb [hunk 2]
```diff
@@ -23,6 +25,29 @@ def process_fields(record)
...
+
+            csv = CSV.read(csv_files[0], headers: true)
+            brca1_count, apc_count, mlh1_count = get_csv_counts(csv)
+            mlh1_count + apc_count < brca1_count
+          end
+
+          def get_csv_counts(csv)
+            brca1_count = csv['mapped:gene'].tally['7'].to_i
+            apc_count = csv['mapped:gene'].tally['358'].to_i
+            mlh1_count = csv['mapped:gene'].tally['2744'].to_i
```
Comment (ollietulloch) [inline lib/import/brca/providers/oxford/oxford_handler.rb:47]: I don't think we need to tally `csv['mapped:gene']` 3 times here:

```
gene_tally  = csv['mapped:gene'].tally
brca1_count = gene_tally['7'].to_i
... etc ....
```

File: lib/import/colorectal/providers/oxford/oxford_handler_colorectal.rb [hunk 3]
```diff
@@ -22,6 +24,35 @@ def process_fields(record)
...
+            brca1_count, apc_count, mlh1_count = get_csv_counts(csv)
+
+            mlh1_count + apc_count > brca1_count
+          end
+
+          def get_csv_counts(csv)
+            brca1_count = csv['mapped:gene'].tally['7'].to_i
+            apc_count = csv['mapped:gene'].tally['358'].to_i
+            mlh1_count = csv['mapped:gene'].tally['2744'].to_i
+            [brca1_count, apc_count, mlh1_count]
```
Comment (ollietulloch) [inline lib/import/colorectal/providers/oxford/oxford_handler_colorectal.rb:48]: Similar to [comment](https://github.com/NHSDigital/data_management_system/pull/56/files#r1470077123) above

**Review by shilpigoeldev — CHANGES_REQUESTED** [review]

File: lib/import/colorectal/providers/oxford/oxford_handler_colorectal.rb (hunk 3 above)
Comment (lauramccluskey1) [reply lib/import/colorectal/providers/oxford/oxford_handler_colorectal.rb:49]: See commit [fc77d32](https://github.com/NHSDigital/data_management_system/pull/56/commits/fc77d32297750f5ad5e7af05a8448808d070bffa)

File: lib/import/brca/providers/oxford/oxford_handler.rb (hunk 2 above)
Comment (ollietulloch) [reply lib/import/brca/providers/oxford/oxford_handler.rb]: Thanks 👍

File: lib/import/brca/providers/oxford/oxford_handler.rb [hunk 4]
```diff
@@ -105,8 +130,8 @@ def process_gene(genotype, record)
             genotypes = []
             gene      = record.mapped_fields['gene'].to_i
             synonym   = record.raw_fields['sinonym'].to_s
-            if [7, 8, 79, 865, 3186, 2744, 2804, 3394, 62, 76,
-                590, 3615, 3616, 20, 18].include? gene
+            if [7, 8, 590, 18, 20, 865, 2744, 2804, 2808, 3186, 3394, 62, 3615, 3616, 76, 79, 358,
```
[the remark this answers is not in the document]
Comment (ollietulloch) [reply lib/import/brca/providers/oxford/oxford_handler.rb:134]: I'm not too keen on these lists being a a collection of meaningless PKs. Pragmatically, it's probably fine, PKs are unlikely the change. It does mean the test fixture PKs will need to be aligned, which isn't ideal.

[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/brca/providers/oxford/oxford_handler.rb:134]: Totally agree with this, its also very easy to miss one when its a meaningless list!

[the remark this answers is not in the document]
Comment (shilpigoeldev) [reply lib/import/brca/providers/oxford/oxford_handler.rb:134]: ok

[the remark this answers is not in the document]
Comment (ollietulloch) [reply lib/import/brca/providers/oxford/oxford_handler.rb:134]: could it be `Zgene.where(shortdesc: CONSTANT_OF_SHORTDESCS).pluck(:zgeneid)`?

[the remark this answers is not in the document]
Comment (shilpigoeldev) [reply lib/import/brca/providers/oxford/oxford_handler.rb:134]: @ollietulloch That would have been the ideal thing to do , but unfortunately we don't have Zgene table available in DMS db..

[the remark this answers is not in the document]
Comment (ollietulloch) [reply lib/import/brca/providers/oxford/oxford_handler.rb:134]: Ah ok, where are the list of Zgene ids coming from?

[the remark this answers is not in the document]
Comment (shilpigoeldev) [reply lib/import/brca/providers/oxford/oxford_handler.rb:134]: We refer Zgene table of Encore.

[the remark this answers is not in the document]
Comment (ollietulloch) [reply lib/import/brca/providers/oxford/oxford_handler.rb:134]: Ok, there isn't much we can do here then - we'll just have to maintain the kist of PKs. When this code moves to encore, we can use the zgene table.

@lauramccluskey1 - could you add a TODO comment above the code please? Something like `# TODO: reference the Zgene table once this code is moved to Era`.

File: lib/import/colorectal/providers/oxford/oxford_handler_colorectal.rb [hunk 5]
```diff
@@ -22,6 +24,35 @@ def process_fields(record)
...
+          end
+
+          def convert_nil_to_zero(count)
+            count = 0 if count.nil?
+            count
+          end
```
[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/colorectal/providers/oxford/oxford_handler_colorectal.rb]: See commit [79a0998](https://github.com/NHSDigital/data_management_system/pull/56/commits/79a0998732bc539f372b1be90b9e8b58fa25308c)

File: lib/import/brca/providers/oxford/oxford_handler.rb [hunk 6]
```diff
@@ -140,14 +165,17 @@ def full_screen?(record)
             return false if record.raw_fields['scope / limitations of test'].nil?
 
             geneticscope = record.raw_fields['scope / limitations of test']
-            geneticscope.scan(/panel|scree(n|m)|brca|hcs|panel/i).size.positive?
+            geneticscope.scan(/panel|full\s?screen|full\sscreem|full\sgene\sscreen|
+                              brca_multiplicom|hcs|brca1|brca2|CNV.*only|CNV.*analysis|
```
[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/brca/providers/oxford/oxford_handler.rb]: See commit [79a0998](https://github.com/NHSDigital/data_management_system/pull/56/commits/79a0998732bc539f372b1be90b9e8b58fa25308c)

File: lib/import/brca/providers/oxford/oxford_handler.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (lauramccluskey1) [reply lib/import/brca/providers/oxford/oxford_handler.rb]: See commit [79a0998](https://github.com/NHSDigital/data_management_system/pull/56/commits/79a0998732bc539f372b1be90b9e8b58fa25308c)

File: lib/import/helpers/brca/providers/rth/constants.rb [hunk 7]
```diff
@@ -64,14 +64,10 @@ module Constants
...
-                         exons?\s?(?<location>[0-9]+(?:-[0-9]+)?)|
-                         ex(?:on)?\s?(?<location>[0-9]+(?:-[0-9]+)?)\s?(?<variant>del|inv|dup)|
-                         exons?\s?(?<location>[0-9]+(?:-[0-9]+)?)\s?(?<variant>del|inv|dup)|
-                         ex\s?(?<location>[0-9]+\+[0-9]+)\s?(?<variant>del|inv|dup)|
-                         (?<variant>del|inv|dup).+ion\sBRCA1\sexons?\s(?<location>[0-9]+-(?:[0-9]*))|
-                         exon\s?(?<location>[0-9]+)\s?-exon\s?(?<moreexon>[0-9]+)
-                         \s?(?<variant>del|inv|dup)/ix
+            EXON_REGEX = /(?<mutationtype>del|inv|dup).+ion\s(?:of\s)?exons?\s?(?<exons>[0-9]+(?:-[0-9]+)?)|
+                          ex(?:on|ons)?\s?(?<exons>[0-9]+(?:(?:-|\+)[0-9]+)?)\s?(?<mutationtype>del|inv|dup)|
+                          (?<mutationtype>del|inv|dup).+ion\sBRCA1\sexons?\s(?<exons>[0-9]+-(?:[0-9]*))|
```
Comment (shilpigoeldev) [inline lib/import/helpers/brca/providers/rth/constants.rb:69]: Can it be made more generic ex- _[a-z ]+_ rather _BRCA1_ to be able to extract mutation and exonic location irrespective of what comes ex- BRCA2 rather BRCA1?

**Review by shilpigoeldev — CHANGES_REQUESTED** [review]

File: lib/import/helpers/brca/providers/rth/constants.rb (hunk 7 above)
Comment (lauramccluskey1) [reply lib/import/helpers/brca/providers/rth/constants.rb]: See commit [a76e454](https://github.com/NHSDigital/data_management_system/pull/56/commits/a76e45453449614aceec720cd7730fb4d5d7efc9) and [9c4ffb7](https://github.com/NHSDigital/data_management_system/pull/56/commits/9c4ffb747ead4916222b1cb06408c269bc9e2d7a)

**Comment by lauramccluskey1** [comment]
@shilpigoeldev- the variant counts are the same after the changes for both brca and colorectal

**Review by shilpigoeldev — APPROVED** [review]
Thanks @lauramccluskey1 ..looks good to me.. Happy to approve if the counts are matching and QAed by Fiona.

**Pull request merged**